### PR TITLE
Nectar cam reader

### PR DESCRIPTION
**Pull request by FrancaCassol**
First version of the a eventsource for NectarCAM data. 
Only a minimal part of the specific camera data are included for the moment.

### CODE REVIEW

**Comment by dneise** [comment]
It looks like four empty packages have been created in this PR, by creating these empty `__init__.py`-files.
```
dneise@host:ctapipe/atmosphere$ tree
├── fram
│   └── __init__.py
├── gdas
│   └── __init__.py
├── __init__.py
└── lidar
    └── __init__.py
```

Did this maybe happen by accident or will these package be filled with functionality in this PR?

File: ctapipe/io/tests/test_nectarcameventsource.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    assert i == N_EVENTS - 1
+
+
+def test_is_compatible():
+    from ctapipe.io.nectarcameventsource import NectarCAMEventSource
+
+    assert NectarCAMEventSource.is_compatible(example_file_path)
+
+
+def test_factory_for_protozfits_file():
```
Comment (dneise) [inline ctapipe/io/tests/test_nectarcameventsource.py]: I think this function name should maybe be "test_factory_for_nectarcam_file". Accordingly I should rename the test function analog to this function for the sst1m_event_source.

File: docs/atmosphere/index.rst [hunk 2]
```diff
@@ -0,0 +1,47 @@
+.. _atmosphere:
+
+==========================
+ Atmosphere (`atmosphere`)
```
Comment (dneise) [inline docs/atmosphere/index.rst]: This PR seems to be about nectarcam event source ... but some things related to the atmosphere seem to have somehow found their way into this PR... 

Maybe it is better to remove this for the moment from this PR and open a dedicated "Atmosphere" PR

File: ctapipe/io/tests/test_nectarcameventsource.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+from pkg_resources import resource_filename
+import os
+
+import pytest
+pytest.importorskip("protozfits", minversion="0.44.3")
```
Comment (dneise) [inline ctapipe/io/tests/test_nectarcameventsource.py]: I think this should be `minversion="0.44.4"`

**Comment by dneise** [comment]
Ah ... I just realize ... I think you need v0.44.4 of the `protozfits` for this .. since only v0.44.4 contains the example file you need for testing. 

So you need to modify the  `.travis.yml` file to install this particular version, here:
https://github.com/cta-observatory/ctapipe/blob/e4d9a307ec7bf2846bc5930de9b175ee9f5cf41a/.travis.yml#L57

File: docs/atmosphere/index.rst (hunk 2 above)
Comment (FrancaCassol) [reply docs/atmosphere/index.rst]: You are right, strange, it should have been pulled already some time ago. How can take it back from this PR?

**Comment by dneise** [comment]
Ah ... nearly forgot: Thank you for this PR, I hope this might be useful to find out more about similarities and differences of all the different telescopes.

**Comment by dneise** [comment]
Tests are now failing in one seemingly unrelated place. Not sure why ... digging into it...
https://travis-ci.org/cta-observatory/ctapipe/jobs/355924744#L1485

File: docs/atmosphere/index.rst (hunk 2 above)
Comment (dneise) [reply docs/atmosphere/index.rst]: there are multiple ways:

### simple
just `git rm` these 4 `__init__.py` files and the `index.rst` file that were accidentally added.

git tries to make sure we never loose, when ever we have written once, so this "adding-and-removing" will be part of the history of this branch and also later part of the history of the master. Maybe not ideal, but imho also no big deal.

### cherry-pick

Make a fresh branch from the master (after making sure your fork is up-to-date) and give it a slightly different name and `git cherry-pick` only those commits from this branch, which you actually want to have in a commit.

### ... more

There are more ways ... I have never tried....

Comment (dneise) [reply docs/atmosphere/index.rst]: opened a little PR for this here:
https://github.com/FrancaCassol/ctapipe/pull/2

**Comment by FrancaCassol** [comment]
Hi @kosack , my pull seems to be blocked by the problem with the test_targetio_event_source.py (that I didn't touch), can you please help me to head with the merge?

**Comment by watsonjj** [comment]
```
    def test_eventsourcefactory():
        dataset = get_dataset("chec_r1.tio")
>       source = EventSourceFactory.produce(input_url=dataset)
ctapipe/io/tests/test_targetio_event_source.py:126: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ctapipe/core/factory.py:245: in produce
    instance = factory._instance
ctapipe/core/factory.py:194: in _instance
    product = self._product
ctapipe/core/factory.py:171: in _product
    product_name = self._get_product_name()
ctapipe/io/eventsourcefactory.py:64: in _get_product_name
    if subclass.is_compatible(self.input_url):
ctapipe/io/nectarcameventsource.py:45: in is_compatible
    h = fits.open(file_path)[2].header
../../../miniconda/envs/cta-dev/lib/python3.6/site-packages/astropy/io/fits/hdu/hdulist.py:317: in __getitem__
    raise e
../../../miniconda/envs/cta-dev/lib/python3.6/site-packages/astropy/io/fits/hdu/hdulist.py:313: in __getitem__
    self._positive_index_of(key))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = [<astropy.io.fits.hdu.image.PrimaryHDU object at 0x7f54636b5f60>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x7f54636b09e8>]
func = <built-in method __getitem__ of HDUList object at 0x7f5463541e08>
args = (2,), kwargs = {}
    def _try_while_unread_hdus(self, func, *args, **kwargs):
        """
            Attempt an operation that accesses an HDU by index/name
            that can fail if not all HDUs have been read yet.  Keep
            reading HDUs until the operation succeeds or there are no
            more HDUs to read.
            """
    
        while True:
            try:
>               return func(*args, **kwargs)
E               IndexError: list index out of range
```

The problem is there is an IndexError when applying NectarCAMEventSource.is_compatible to a CHEC data file, as technically the file format used by CHEC is a fits file, but not of a format astropy.fits expects.

In NectarCAMEventSource.is_compatible add the last 3 lines of this block:
```python
        try:
            # The file contains two tables:
            #  1: RunHeader
            #  2: Events <--- this is what we need to look at
            h = fits.open(file_path)[2].header
            ttypes = [
                h[x] for x in h.keys() if 'TTYPE' in x
            ]
        except OSError:
            # not even a fits file
            return False
        except IndexError:
            # A fits file of a different format
            return False
```

**Comment by FrancaCassol** [comment]
Hi Karl,

I am a bit lost, why did this pull request not work?

**Comment by kosack** [comment]
>I am a bit lost, why did this pull request not work? 

It seems fine - I just didn't have a time to do a code review yet :-) I'll do it now!

**Comment by FrancaCassol** [comment]
Thanks! 

F.

**Pull request merged**